playbooks/upload_to_swift.yml

```diff
@@ -62,23 +101,107 @@
+    # So they don't show up on the final page.
+    - name: Generate file list with sizes for index page data
+      shell: |
+        find . \
+          -type f \
+          -mindepth 1 \
+          -exec ls -l {} \; \
+        |sed 's+\./++g'
+      args:
+        chdir: "{{ adir }}"
+      register: file_list
+
+    - name: Copy index.html
+      copy:
+        src: templates/artifact/index.html
+        dest: "{{ adir }}/index.html"
+
+    - name: Copy style.css
+      copy:
+        src: templates/artifact/styles.css
+        dest: "{{ adir }}/styles.css"
+
+    # this is json data containing metadata and a list of files
+    # that will be downloaded by index.html
+    # and used to display a list of available artifacts
+    - name: Generate data.json
+      copy:
+        content: |
+          {
+            "job_name": "{{ job_name }}",
+            "build_number": "{{ build_number }}",
+            "archive_base_name": "{{ archive_base_name }}",
+            "container_public_url": "{{ container_public_url}}",
+            "files": [
+              {% for file_line in file_list.stdout_lines %}
+                {% set file_list = file_line.split() %}
+                {% if file_list|length > 3 %}
+                  {
+                    "path": "{{file_list[-1]}}",
+                    "size": {{file_list[4]}}
+                  }{{ "," if not loop.last else "" }}
+                {% endif %}
+              {% endfor %}
+            ],
+            "archives": [
+              "{{ container_public_url }}/{{ archive_base_name }}.tar.gz"
+            ]
+          }
+        dest: "{{ adir }}/data.json"
+
+    # os_object does not currently support setting object expiration header field
+    # production retention
+    - name: Upload Artifacts to Cloud Files
+      command: "swift upload --object-threads 100 --ignore-checksum --header 'X-Delete-After:{{ retention }}' {{ container }} {{ adir | basename }}"
+      args:
+        chdir: "{{ adir |dirname }}"
+      environment:
+        OS_AUTH_TOKEN: "{{ auth_token }}"
+        OS_STORAGE_URL: "{{ (object_store['endpoints'] | selectattr('region', 'equalto', region) | first)['publicURL'] }}"
+      failed_when: false
+
+    - name: Wait for async archive creation to complete
+      async_status:
+        jid: "{{ create_archive_async.ansible_job_id }}"
+      register: caa
+      until: caa.finished
+      retries: 180
+      delay: 60
+
+    # os_object does not currently support setting object expiration header field
+    - name: Upload archive and index files to Cloud Files
+      command: >-
+        swift upload {{ container }} {{ archive_base_name }}.tar.gz
+        --header 'X-Delete-After: {{ retention }}'
+      args:
+        chdir: "{{ adir | dirname }}"
+      environment:
+        OS_AUTH_TOKEN: "{{ auth_token }}"
+        OS_STORAGE_URL: "{{ (object_store['endpoints'] | selectattr('region', 'equalto', region) | first)['publicURL'] }}"
```

Comment:
Should we add retries here?

Reply:
done

Reply:
Added a section for upload failures, that only appears if there is at leat one failure: https://901a6f3144ea1698911a-fb37949d4321c68f6e404c5dda72c3d1.ssl.cf5.rackcdn.com/hstest_123/index.html